server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java

```diff
@@ -150,19 +153,28 @@ final class DefaultSearchContext extends SearchContext {
         this.indexShard = readerContext.indexShard();
 
         Engine.Searcher engineSearcher = readerContext.acquireSearcher("search");
-        this.searcher = new ContextIndexSearcher(
-            engineSearcher.getIndexReader(),
-            engineSearcher.getSimilarity(),
-            engineSearcher.getQueryCache(),
-            engineSearcher.getQueryCachingPolicy(),
-            minimumDocsPerSlice,
-            lowLevelCancellation,
-            // TODO not set the for now, this needs a special thread pool and can be enabled after its introduction
-            // parallelize
-            // ? (EsThreadPoolExecutor) this.indexService.getThreadPool().executor(ThreadPool.Names.CONCURRENT_COLLECTION_TBD)
-            // : null,
-            null
-        );
+        if (executor == null) {
+            this.searcher = new ContextIndexSearcher(
+                engineSearcher.getIndexReader(),
+                engineSearcher.getSimilarity(),
+                engineSearcher.getQueryCache(),
+                engineSearcher.getQueryCachingPolicy(),
+                lowLevelCancellation
+            );
+        } else {
+            this.searcher = new ContextIndexSearcher(
+                engineSearcher.getIndexReader(),
+                engineSearcher.getSimilarity(),
+                engineSearcher.getQueryCache(),
+                engineSearcher.getQueryCachingPolicy(),
+                lowLevelCancellation,
+                executor,
+                forceSequentialCollection,
+                minimumDocsPerSlice,
+                // TODO is there a better way that does not require casting the executor?
+                ((ThreadPoolExecutor) executor).getMaximumPoolSize()
```

Comment:
this cast is not great. We had it before, and I have made it effective only when concurrency is enabled. There was SnapshotResiliencyTests that injects a different executor and was causing a class cast exception here, to I went for disabling concurrency in that test entirely via the cluster setting.

We could also expose the max pool size (searchOrGetThreadPoolSize) in some way in ThreadPool instead of retrieving it from the thread pool executor, but I am not too sure that that is better.